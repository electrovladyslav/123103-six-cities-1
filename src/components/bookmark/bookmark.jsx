import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Link} from "react-router-dom";
import {connect} from "react-redux";

import {getAuthrizationStatus} from "../../selectors";
import {Operation} from "../../reducer";
import {BookmarkSize, RentCardClassesEnum} from "../../constants";

class Bookmark extends PureComponent {
  constructor(props) {
    super(props);

    this._handleClick = this._handleClick.bind(this);
    this.bookmarkClass = props.bookmarkClass || RentCardClassesEnum.RENT_CARD;
  }

  render() {
    const {isFavorite, isAuthorized} = this.props;
    return (
      <button
        className={`${this.bookmarkClass}__bookmark-button button ${
          isFavorite ? `${this.bookmarkClass}__bookmark-button--active` : ``
        }`}
        type="button"
        onClick={this._handleClick}>
        {this._renderFlag(isAuthorized)}
      </button>
    );
  }

  _handleClick(event) {
    if (this.props.isAuthorized) {
      const isAdding = this.props.isFavorite ? 0 : 1;
      this.props
        .postToFavorites(isAdding, this.props.offerId)
        .then(this.props.toggleFavoriteStatus(!!isAdding));
    }
    event.preventDefault();
  }

  _renderFlag(isAuthorized) {
    const bookmarkSize = this.props.bookmarkSize || BookmarkSize.NORMAL;

    if (isAuthorized) {
      return (
        <React.Fragment>
          <svg
            className={`${this.bookmarkClass}__bookmark-icon`}
            width={bookmarkSize.width}
            height={bookmarkSize.height}>
            <use xlinkHref="#icon-bookmark" />
          </svg>
          <span className="visually-hidden">To bookmarks</span>
        </React.Fragment>
      );
    } else {
      return (
        <Link to="/login">
          <svg
            className={`${this.bookmarkClass}__bookmark-icon`}
            width={bookmarkSize.width}
            height={bookmarkSize.height}>
            <use xlinkHref="#icon-bookmark" />
          </svg>
          <span className="visually-hidden">To bookmarks</span>
        </Link>
      );
    }
  }
}

Bookmark.propTypes = {
  isFavorite: PropTypes.bool.isRequired,
  isAuthorized: PropTypes.bool,
  bookmarkSize: PropTypes.shape({
    width: PropTypes.number,
    height: PropTypes.number,
  }),
  bookmarkClass: PropTypes.string,
  offerId: PropTypes.number,
  postToFavorites: PropTypes.func,
  toggleFavoriteStatus: PropTypes.func,
};

const mapStateToProps = (state, ownProps) =>
  Object.assign({}, ownProps, {
    isAuthorized: getAuthrizationStatus(state),
  });

const mapDispatchToProps = (dispatch) => ({
  postToFavorites: (status, offerId) =>
    dispatch(Operation.postToFavorites(status, offerId)),
});

export {Bookmark};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Bookmark);
