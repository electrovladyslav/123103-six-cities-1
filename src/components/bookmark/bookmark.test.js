import React from "react";
import renderer from "react-test-renderer";
import {StaticRouter} from "react-router";

import {Bookmark} from "./bookmark.jsx";

it(`Bookmark component renders correctly`, () => {
  const tree = renderer
    .create(
        <StaticRouter>
          <Bookmark isFavorite={false} offerId={42} />
        </StaticRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
