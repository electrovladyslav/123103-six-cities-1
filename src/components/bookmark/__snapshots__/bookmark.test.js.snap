// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark component renders correctly 1`] = `
<button
  className="place-card__bookmark-button button "
  onClick={[Function]}
  type="button"
>
  <a
    href="/login"
    onClick={[Function]}
  >
    <svg
      className="place-card__bookmark-icon"
      height={19}
      width={18}
    >
      <use
        xlinkHref="#icon-bookmark"
      />
    </svg>
    <span
      className="visually-hidden"
    >
      To bookmarks
    </span>
  </a>
</button>
`;
